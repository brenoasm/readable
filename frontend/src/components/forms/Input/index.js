import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';

import FormErrorMessage from '../FormErrorMessage';

const defaultProps = {
  name: '',
  title: '',
  className: '',
  type: 'text',
  value: '',
  id: '',
  handleChange: () => {},
  placeholder: '',
  errors: []
};

const propTypes = {
  name: PropTypes.string,
  title: PropTypes.string,
  className: PropTypes.string,
  type: PropTypes.string,
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number
  ]),
  id: PropTypes.string,
  handleChange: PropTypes.func,
  placeholder: PropTypes.string,
  errors: PropTypes.arrayOf(PropTypes.string)
};

const StyledInput = styled.div`
  margin: 5px;

  & > label {
    margin-right: 5px;
  }

  & > input {
    border: 2px solid rgb(238, 238, 238);
  }
`;

const Input = ({
  name,
  title,
  className,
  type,
  value,
  id,
  handleChange,
  placeholder,
  errors
}) => {
  return (
    <StyledInput>
      <label htmlFor={name}>{title}</label>
      <input
        className={className}
        type={type}
        value={value}
        id={id}
        name={name}
        onChange={handleChange}
        placeholder={placeholder}
      />
      <FormErrorMessage errors={errors} />
    </StyledInput>
  )
};

Input.propTypes = propTypes;
Input.defaultProps = defaultProps;

export default Input;
