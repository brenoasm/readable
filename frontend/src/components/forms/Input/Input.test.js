/* eslint-disable */

import React from 'react';

import Input from './';

describe('Input', () => {
  it('matches the snapshot', () => {
    const initialState = {
      name: 'author',
      title: 'author',
      type: 'text',
      value: 'author',
      id: 'author',
      handleChange: jest.fn(),
      placeholder: 'Testando o Input',
      errors: []
    };

    const wrapper = shallow(<Input {...initialState}/>);

    expect(wrapper).toMatchSnapshot();
  });

  it('shows a title when the property is passed', () => {
    const initialState = {
      name: 'author',
      title: 'author',
      type: 'text',
      value: 'author',
      id: 'author',
      handleChange: jest.fn(),
      placeholder: 'Testando o Input',
      errors: []
    };

    const wrapper = shallow(<Input {...initialState}/>);

    expect(wrapper.find('label').text()).toBe(initialState.title);
  });

  it('should have one FormErrorMessage', () => {
    const initialState = {
      errors: [
        'Erro ao inserir valor'
      ]
    };

    const wrapper = shallow(<Input {...initialState}/>);

    expect(wrapper.find('FormErrorMessage')).toHaveLength(1);
  });

  it('trigger the handleChange event when the value changes', () => {
    const initialState = {
      handleChange: jest.fn(),
    };

    const wrapper = shallow(<Input {...initialState}/>);

    wrapper.find('input').simulate('change', { target: { value: 'a' }});

    expect(initialState.handleChange).toHaveBeenCalled();
  });

  it('adapts to the parameter Type', () => {
    const initialState = {
      handleChange: jest.fn(),
      type: 'checkbox'
    };

    const wrapper = shallow(<Input {...initialState}/>);

    expect(wrapper.find('input').prop('type')).toBe('checkbox');
  });
});
