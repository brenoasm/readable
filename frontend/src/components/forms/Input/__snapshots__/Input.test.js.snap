// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input matches the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Input
    className=""
    errors={Array []}
    handleChange={[Function]}
    id="author"
    name="author"
    placeholder="Testando o Input"
    title="author"
    type="text"
    value="author"
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <label
          htmlFor="author"
>
          author
</label>,
        <input
          className=""
          id="author"
          name="author"
          onChange={[Function]}
          placeholder="Testando o Input"
          type="text"
          value="author"
/>,
        <FormErrorMessage
          errors={Array []}
/>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "author",
          "htmlFor": "author",
        },
        "ref": null,
        "rendered": "author",
        "type": "label",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "",
          "id": "author",
          "name": "author",
          "onChange": [Function],
          "placeholder": "Testando o Input",
          "type": "text",
          "value": "author",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "errors": Array [],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <label
            htmlFor="author"
>
            author
</label>,
          <input
            className=""
            id="author"
            name="author"
            onChange={[Function]}
            placeholder="Testando o Input"
            type="text"
            value="author"
/>,
          <FormErrorMessage
            errors={Array []}
/>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "author",
            "htmlFor": "author",
          },
          "ref": null,
          "rendered": "author",
          "type": "label",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "",
            "id": "author",
            "name": "author",
            "onChange": [Function],
            "placeholder": "Testando o Input",
            "type": "text",
            "value": "author",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "errors": Array [],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
